File: test/language/enumeration_types/cpp98/VarUIntEnumTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"
#include "zserio/CppRuntimeException.h"

#include "enumeration_types/varuint_enum/DarkColor.h"

namespace enumeration_types
{
namespace varuint_enum
{

class VarUIntEnumTest : public ::testing::Test
{
protected:
    static const size_t DARK_COLOR_NONE_BITSIZEOF = 8;
    static const size_t DARK_COLOR_DARK_GREEN_BITSIZEOF = 16;
};

TEST_F(VarUIntEnumTest, emptyConstructor)
{
    const DarkColor darkColor;
    ASSERT_EQ(0, darkColor.getValue());
}

TEST_F(VarUIntEnumTest, valueConstructor)
{
    const DarkColor darkColor(DarkColor::DARK_RED);
    ASSERT_EQ(1, darkColor.getValue());
}

TEST_F(VarUIntEnumTest, bitStreamReaderConstructor)
{
    zserio::BitStreamWriter writer;
    writer.writeVarUInt(static_cast<uint32_t>(DarkColor::DARK_GREEN));
    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);

    const DarkColor darkColor(reader);
    ASSERT_EQ(255, darkColor.getValue());
}

TEST_F(VarUIntEnumTest, operatorFunction)
{
    const DarkColor darkColor(DarkColor::DARK_BLUE);
    const DarkColor::e_DarkColor colorEnum = darkColor;
    ASSERT_EQ(DarkColor::DARK_BLUE, colorEnum);
}

TEST_F(VarUIntEnumTest, getValue)
{
    const DarkColor darkColor(DarkColor::DARK_GREEN);
    ASSERT_EQ(255, darkColor.getValue());
}

TEST_F(VarUIntEnumTest, bitSizeOf)
{
    const DarkColor darkColor1(DarkColor::NONE);
    ASSERT_TRUE(darkColor1.bitSizeOf() == DARK_COLOR_NONE_BITSIZEOF);

    const DarkColor darkColor2(DarkColor::DARK_GREEN);
    ASSERT_TRUE(darkColor2.bitSizeOf() == DARK_COLOR_DARK_GREEN_BITSIZEOF);
}

TEST_F(VarUIntEnumTest, initializeOffsets)
{
    const size_t bitPosition = 1;
    const DarkColor darkColor1(DarkColor::NONE);
    ASSERT_TRUE(darkColor1.initializeOffsets(bitPosition) == bitPosition + DARK_COLOR_NONE_BITSIZEOF);

    const DarkColor darkColor2(DarkColor::DARK_GREEN);
    ASSERT_TRUE(darkColor2.initializeOffsets(bitPosition) == bitPosition + DARK_COLOR_DARK_GREEN_BITSIZEOF);
}

TEST_F(VarUIntEnumTest, operatorEquality)
{
    const DarkColor color1(DarkColor::DARK_RED);
    const DarkColor color2;
    const DarkColor color3(DarkColor::DARK_GREEN);
    const DarkColor color4(DarkColor::DARK_RED);

    ASSERT_FALSE(color1 == color2);
    ASSERT_FALSE(color1 == color3);
    ASSERT_TRUE(color1 == color4);
}

TEST_F(VarUIntEnumTest, operatorEnumEquality)
{
    const DarkColor color1(DarkColor::DARK_RED);
    ASSERT_FALSE(color1 == DarkColor::DARK_GREEN);
    ASSERT_TRUE(color1 == DarkColor::DARK_RED);
}

TEST_F(VarUIntEnumTest, hashCode)
{
    const DarkColor color1(DarkColor::DARK_RED);
    const DarkColor color2;
    const DarkColor color3(DarkColor::DARK_GREEN);
    const DarkColor color4(DarkColor::DARK_RED);

    ASSERT_NE(color1.hashCode(), color2.hashCode());
    ASSERT_NE(color1.hashCode(), color3.hashCode());
    ASSERT_EQ(color1.hashCode(), color4.hashCode());
}

TEST_F(VarUIntEnumTest, read)
{
    zserio::BitStreamWriter writer;
    writer.writeVarUInt(static_cast<uint32_t>(DarkColor::DARK_RED));
    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);

    DarkColor darkColor;
    ASSERT_NE(1, darkColor.getValue());
    darkColor.read(reader);
    ASSERT_EQ(1, darkColor.getValue());
}

TEST_F(VarUIntEnumTest, write)
{
    const DarkColor darkColor(DarkColor::DARK_BLUE);
    zserio::BitStreamWriter writer;
    darkColor.write(writer);

    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);
    ASSERT_EQ(2, reader.readVarUInt());
}

TEST_F(VarUIntEnumTest, toString)
{
    ASSERT_EQ(std::string("NONE"), std::string(DarkColor(DarkColor::NONE).toString()));
    ASSERT_EQ(std::string("DARK_RED"), std::string(DarkColor(DarkColor::DARK_RED).toString()));
    ASSERT_EQ(std::string("DARK_BLUE"), std::string(DarkColor(DarkColor::DARK_BLUE).toString()));
    ASSERT_EQ(std::string("DARK_GREEN"), std::string(DarkColor(DarkColor::DARK_GREEN).toString()));
}

TEST_F(VarUIntEnumTest, toEnum)
{
    ASSERT_EQ(DarkColor::NONE, DarkColor::toEnum(0));
    ASSERT_EQ(DarkColor::DARK_RED, DarkColor::toEnum(1));
    ASSERT_EQ(DarkColor::DARK_BLUE, DarkColor::toEnum(2));
    ASSERT_EQ(DarkColor::DARK_GREEN, DarkColor::toEnum(255));
}

TEST_F(VarUIntEnumTest, toEnumFailure)
{
    ASSERT_THROW(DarkColor::toEnum(3), zserio::CppRuntimeException);
}

} // namespace varuint_enum
} // namespace enumeration_types

File: test/language/enumeration_types/java/enumeration_types/varuint_enum/VarUIntEnumTest.java
package enumeration_types.varuint_enum;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.math.BigInteger;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class VarUIntEnumTest
{
    @Test
    public void constructor()
    {
        final DarkColor darkColor = DarkColor.DARK_RED;
        assertEquals(DarkColor.DARK_RED, darkColor);
    }

    @Test
    public void getValue()
    {
        final DarkColor darkColor = DarkColor.DARK_BLUE;
        assertEquals(DARK_BLUE_VALUE, darkColor.getValue());
    }

    @Test
    public void getGenericValue()
    {
        final DarkColor darkColor = DarkColor.DARK_GREEN;
        assertEquals(DARK_GREEN_VALUE, darkColor.getGenericValue());
    }

    @Test
    public void bitSizeOf()
    {
        final DarkColor darkColor1 = DarkColor.NONE;
        assertEquals(DARK_COLOR_NONE_BITSIZEOF, darkColor1.bitSizeOf());

        final DarkColor darkColor2 = DarkColor.DARK_RED;
        assertEquals(DARK_COLOR_DARK_RED_BITSIZEOF, darkColor2.bitSizeOf());

        final DarkColor darkColor3 = DarkColor.DARK_BLUE;
        assertEquals(DARK_COLOR_DARK_BLUE_BITSIZEOF, darkColor3.bitSizeOf());

        final DarkColor darkColor4 = DarkColor.DARK_GREEN;
        assertEquals(DARK_COLOR_DARK_GREEN_BITSIZEOF, darkColor4.bitSizeOf());
    }

    @Test
    public void write() throws IOException
    {
        final DarkColor darkColor = DarkColor.DARK_GREEN;
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        darkColor.write(writer);
        final BitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray());
        final BigInteger readColor = reader.readVarUInt();
        assertEquals(readColor, darkColor.getValue());
    }

    @Test
    public void toEnum()
    {
        DarkColor darkColor = DarkColor.toEnum(NONE_VALUE);
        assertEquals(DarkColor.NONE, darkColor);

        darkColor = DarkColor.toEnum(DARK_RED_VALUE);
        assertEquals(DarkColor.DARK_RED, darkColor);

        darkColor = DarkColor.toEnum(DARK_BLUE_VALUE);
        assertEquals(DarkColor.DARK_BLUE, darkColor);

        darkColor = DarkColor.toEnum(DARK_GREEN_VALUE);
        assertEquals(DarkColor.DARK_GREEN, darkColor);
    }

    @Test
    public void toEnumFailure()
    {
        assertThrows(IllegalArgumentException.class, () -> DarkColor.toEnum(new BigInteger("3")));
    }

    private static int DARK_COLOR_NONE_BITSIZEOF = 8;
    private static int DARK_COLOR_DARK_RED_BITSIZEOF = 8;
    private static int DARK_COLOR_DARK_BLUE_BITSIZEOF = 8;
    private static int DARK_COLOR_DARK_GREEN_BITSIZEOF = 16;

    private static BigInteger NONE_VALUE = BigInteger.valueOf(0);
    private static BigInteger DARK_RED_VALUE = BigInteger.valueOf(1);
    private static BigInteger DARK_BLUE_VALUE = BigInteger.valueOf(2);
    private static BigInteger DARK_GREEN_VALUE = BigInteger.valueOf(255);
}
